Split the registry client out so it can be used in more places

smarterclayton opened the pull request:
The client was previously in pkg/image/importer but isn't specific to
import. It's really a factory for working with arbitrary registries for
both push and pull, so move it into its own package and trim its
dependency set down to only docker/distribution, which makes it more
portable to different codebases.

@bparees helps use this code in more places

bparees commented:
+1

soltysh commented on pkg/image/registryclient/credentials.go:
```diff
@@ -0,0 +1,86 @@
+package registryclient
```
You're copy&pasting this (at least parts) from `pkg/image/importer/credentials.go` but I don't see that file (those bits) being removed?

soltysh commented on pkg/image/registryclient/credentials.go:
```diff
@@ -0,0 +1,86 @@
+type noopCredentialStore struct{}
+
+func (s *noopCredentialStore) Basic(url *url.URL) (string, string) {
+	glog.Infof("asked to provide Basic credentials for %s", url)
```
While you're at it please remove these debug statements here and in the next few functions.

mfojtik commented:
@smarterclayton is the intention to use the client in registry as well (I think registry also has its own client...?). If so, don't we want to extract this into separate repo or something that is vendorable?

bparees commented:
> @smarterclayton is the intention to use the client in registry as well (I think registry also has its own client...?). If so, don't we want to extract this into separate repo or something that is vendorable?

image-registry uses this client already (by vendoring origin) as i understand it. So yes, moving it to a shared repository so it's vendorable would be ideal but that's also true for a bunch of other code and i don't think we're in a position to take on that refactor 5 days before feature cut.

smarterclayton commented:
Updated.  I split out more so the dependency tree is a lot clearer.  registry client depends on docker distribution.  registryclient/dockercredentials depends on kubernetes/pkg/credentialprovider.  image/importer has anything kube api dependent.

soltysh commented on pkg/image/registryclient/dockercredentials/credentials_test.go:
```diff
@@ -0,0 +1,84 @@
+	_ "github.com/openshift/origin/pkg/api/install"
+)
+
+type mockKeyring struct {
```
Same as before, please remove these duplicates from `pkg/image/importer/credentials_test.go`

smarterclayton replied:
These are local private types.  They aren't duplicates, since we can't import them.

smarterclayton commented:
Tagging based on not being able to export these things.
